File: LeetCode/DP/KnightProbabilityInChessboard.js
/**
 * @param {number} N
 * @param {number} K
 * @param {number} r
 * @param {number} c
 * @return {number}
 * 
 * @jingjiejiang 13, 2019
 */
var knightProbability = function(N, K, r, c) {
    
  let stepCnts = [...Array(N)].fill(1).map(_ => [...Array(N)].fill(1)); 
  const dirs = [[-1,-2], [-2,-1],[-2,1], [-1,2],
                [1,2], [2,1], [2,-1], [1,-2]];
  for (let time = 0; time < K; time ++) {
    const curStepCnts = [...Array(N)].fill(0).map(_ => [...Array(N)].fill(0));
    for (let row = 0; row < N; row ++) {
      for (let col = 0; col < N; col ++) {
        for (let dir of dirs) {
          let curRow = row + dir[0], curCol = col + dir[1];
          if (curRow < 0 || curRow >= N || curCol < 0 || curCol >= N) continue;
          curStepCnts[row][col] += stepCnts[curRow][curCol];
        }
      }
    }
    stepCnts = curStepCnts;
  }

  return stepCnts[r][c] / Math.pow(8, K);
};

File: LeetCode/DP/KnightProbabilityInChessboard.java
import java.util.Arrays;

/**
 * @jingjiejiang Oct 22, 2019
 */
class KnightProbabilityInChessboard {
    public double knightProbability(int N, int K, int r, int c) {

        int[][] dirs = {{1, 2}, {1, -2}, {2, 1}, {2, -1}, {-1, 2}, {-1, -2}, {-2, 1}, {-2, -1}};
        double[][] stepCnts = new double[N][N];

        for (double[] row : stepCnts) {
            Arrays.fill(row, 1);
        }

        for (int times = 0; times < K; times ++) {
            double[][] curSteps = new double[N][N];
            for (int row = 0; row < N; row ++) {
                for (int col = 0; col < N; col ++) {

                    for (int[] dir : dirs) {
                        int newRow = row + dir[0];
                        int newCol = col + dir[1];
                        if (isLegal(N, newRow, newCol)) curSteps[row][col] += stepCnts[newRow][newCol];
                    }
                }
            }

            stepCnts = curSteps;
        }

        return stepCnts[r][c] / Math.pow(8, K);
    }

    private boolean isLegal(int N, int row, int col) {
        return (row >= 0 && row < N && col >= 0 && col < N);
    }

    public static void main(String[] args) {
        // knightProbability(1, 1, 1 ,1);
        // System.out.println("hello world");
    }
}